### DISCUSSION TITLE
What is the best practice to propagate changes to DataObjects, Assets, Documents to other environments?

### DISCUSSION THREAD

**Discussion by ximAABacon** [pimcore/pimcore#17722] [Q&A / Support]
Sorry if this has been asked before, but all I could find is [this](https://pimcore.com/docs/platform/Pimcore/Deployment/Deployment_Tools#content-migration). I imagine this is something which concerns every Pimcore installation. What I am looking for is a best practice to get (DataObjects, Assets, Documents) changes from development to other stages (like integration, staging, live environments).

2 scenarios:

1. The developer creates / modifies a DataObject (not the class definition) on his development environment. What is the process on getting this change to other developers, integration / staging / live environments?
2. Changes are made directly in the staging / live environment. How do these changes flow back to the development environment or from staging to live?

Examples:
1. in the Cars Demo: a new FilterDefinition, a new car, images for the new car are added by the developer. How do these changes and new DataObjects go to the live deployment?
2. content is changed on the staging / live environment. How do these changes go back to the development environments?

**Comment by AlephLau** [comment]
If you are going to modify assets, objects, or documents, you need to do it directly in production. You can do it in development or staging, but ultimately you have to replicate it in production. The only alternative would be to put production in maintenance mode, update staging with the information from production, do whatever you need to do in staging, and then import the database back to production.

If you want to update development with what you have in staging, as you mentioned in the example, you would have to import the database from staging to development. Keep in mind that every change you make is not only at the code level but also at the database level, and performing partial table or row imports is very cumbersome since objects often have many relationships across different tables. I hope this helps you a bit.

**Reply by ximAABacon** [reply]
Hi @AlephCode thanks for your answer. I think this applies mostly to staging content.

However, i still don't understand how changes in the development are supposed to be deployed to the production environment.

From my example 1: a developer changes (or adds) a DataObject e.g. FilterDefinition on his development machine. How is this then deployed to other environments: test, integration, production?

I imagine one solution is to export everything as files (json, csv, sql, ...), commit to scm and then import everything on the subsequent environments. But the export / import possibilities seem much too limited for this (as far as i can tell so far).

Database dumps are imo out of the question because this would imply huge problems meaning that every developer would need a complete copy of the production DB, all have to be synchronized, and so on.

**Reply by AlephLau** [reply]
Hi @ximAABacon 

First of all, changes like creating assets, documents, or DataObjects should be made directly in one environment, such as production. If you want to update your development and staging environments afterward, you need to export the production database and import it into development and staging, and the developers need to do the same.

The only things that will be updated from development to staging and from staging to production are code changes, such as classes (DataObjects), Object Bricks, Data Importers, etc. However, assets, objects, and documents will never be uploaded via commits.

I hope this clarifies your doubts.

Best regards!
